Add RCTDevSettings module

terribleben opened the pull request:
This decouples non-UI logic from RCTDevMenu into a new module RCTDevSettings.

**Motivation**: This allows developers to change dev settings without depending on the built-in dev menu, e.g. if they want to introduce their own UI, or have other devtools logic that doesn't depend on an action sheet.

It also introduces the RCTDevSettingsDataSource protocol for storing dev tools preferences. This could allow a developer to implement alternative behaviors, e.g. loading the settings from some other config, changing settings based on the user, deciding not to persist some settings, or something else.

The included data source implementation, RCTDevSettingsUserDefaultsDataSource, uses NSUserDefaults and is backwards compatible with the older implementation, so **no workflows or dependent code will break, and old saved settings will persist.**

The RCTDevMenu interface has not changed and is therefore also backwards-compatible, though
some methods are now deprecated.

In order to ensure that RCTDevSettings does not use a string/key interface (feedback from @javache), `RCTDevMenuItem toggleItem...` has been removed and old occurrences of that constructor have been replaced with `buttonItemWithTitleBlock...`.

**Test plan**: Use `react-native init myProject --version exponentjs/react-native#@terribleben/dev-settings` and verify that RCTDevMenu acts the same way as before, including various dev tools. Expectation is that behavior will be unchanged from the previous implementation.

javache commented on React/Modules/RCTDevMenu.h:
```diff
@@ -12,6 +12,12 @@
+ * Is shaking to show the menu enabled. The menu is enabled by default if RCT_DEV=1, but
+ * you may wish to disable it so that you can provide your own shake handler.
+ */
+FOUNDATION_EXPORT NSString * const kRCTDevSettingShakeToShowDevMenu;
```
Use `RCT_EXTERN` instead of `FOUNDATION_EXPORT` for consistency.

javache replied:
Use `RCT_EXTERN` instead of `FOUNDATION_EXPORT` for consistency.

javache commented on React/Modules/RCTDevMenu.m:
```diff
@@ -9,23 +9,15 @@
 
 static NSString *const RCTShowDevMenuNotification = @"RCTShowDevMenuNotification";
-static NSString *const RCTDevMenuSettingsKey = @"RCTDevMenu";
+NSString * const kRCTDevSettingShakeToShowDevMenu = @"shakeToShow";
```
nit: no space between * and const

javache commented on React/Modules/RCTDevMenu.m:
```diff
@@ -228,224 +178,45 @@ - (instancetype)init
                             modifierFlags:UIKeyModifierCommand
                                    action:^(__unused UIKeyCommand *command) {
-                                     weakSelf.executorClass = Nil;
+                                     [weakSelf.bridge.devSettings updateSettingWithValue:@(NO) forKey:kRCTDevSettingIsDebuggingRemotely];
```
I'd prefer a stronger interface on `RCTDevSettings` rather than `updateSettingWithValue:forKey`

nit: use `@NO` instead of `@(NO)`

terribleben replied:
I prefer keeping it weaker (but with predefined key constants) for two reasons:
- Classes can define their own keys, such as `kRCTDevSettingShakeToShowDevMenu` which applies only to the RCTDevMenu UI but can still leverage the generic DevSettings interface.
- The documentation can live by the keys, instead of trying to put it in two places (the key const definition for, e.g. `kRCTDevSettingIsDebuggingRemotely` in addition to the method definition such as `setIsDebuggingRemotely:`)

However, if you disagree, then I can move the RCT_EXPORT'd methods from RCTDevSettings.m into RCTDevSettings.h, providing strongly-defined setters for these properties as you requested.

javache commented on React/Modules/RCTDevMenu.m:
```diff
@@ -495,47 +270,60 @@ - (void)addItem:(RCTDevMenuItem *)item
+      id value = [_bridge.devSettings settingForKey:item.key];
+      if (value != item.value && ![value isEqual:item.value]) {
+        item.value = value;
+        [item callHandler];
```
What behaviour still lives here? We should have a clear separation of responsibility between RCTDevSettings and RCTDevMenu

terribleben replied:
You're right. Will remove this code and change RCTDevMenu to simply call the handler when the item is selected. If developers want to subscribe to the data source changes on their own, they can.

javache commented on React/Modules/RCTDevSettings.h:
```diff
@@ -0,0 +1,92 @@
+
+@protocol RCTDevSettingsDataSource <NSObject>
+
+- (void)updateStoredSettingWithValue:(id)value forKey:(NSString *)key;
```
nit: remove stored from the interface name. Whether a data source decides to store settings or not is an implementation choice.

Please also document these two methods, explaining when they're called / what the expected behaviour is.

javache commented on React/Modules/RCTDevSettings.m:
```diff
@@ -0,0 +1,508 @@
+
+@end
+
+@implementation RCTDevSettings
```
Can you make sure this module is only allocated when RCT_DEV is true?

javache commented on React/Modules/RCTDevSettings.m:
```diff
@@ -0,0 +1,508 @@
+                                               object:_dataSource];
+    
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
```
Why is this `dispatch_once`? Can't we have this logic in the datasource?

terribleben replied:
Good catch, it was just copied over from the old implementation and I overlooked it.

javache commented on React/Modules/RCTDevSettings.m:
```diff
@@ -0,0 +1,508 @@
+    propertiesToUpdate = [notification.userInfo objectForKey:kRCTDevSettingsUpdatedSettingsKey];
+    updateAllProperties = NO;
+  }
+  for (NSString *property in propertiesToUpdate) {
```
This sort of interface is very weakly defined. It's very easy to break something since it's impossible to tell who may be using the API. Avoid using `NSNotification` as an interface.

terribleben replied:
Hmm. Can you give an example of a situation where you're afraid of it breaking? I designed this as a NSNotification so that anybody can subscribe to updates about settings changing in the dataSource.

RCTDevSettings is one such subscriber, because it uses the notification to activate/deactivate its dev tools. If somebody wanted to add a custom dev setting, they would follow the same pattern-- tell the data source to update, subscribe to changes, and take action after a change is posted.

I designed it this way because the old implementation had really convoluted state management, and this seemed like a clean unidirectional model, where there is one authority about the current state of a dev setting (the data source) and potentially many subscribers.

terribleben commented:
Thanks for the feedback @javache and happy holidays!

I added a few comments/questions (some of them got hidden by the outdated diff, sorry) and addressed most other things. I haven't yet responded to "could we "fix" that behaviour here and always use the right executor from the first run?" but I'll think about that next.

terribleben commented:
I put more thought into your question about the bridge executor class. Here's more or less how we would do that:
- We'd remove the `executorClass` logic from RCTDevSettings
- In the `RCTBatchedBridge::executorClass` property, we'd remove the references to `_parentBridge.executorClass` and instead make the getter/setter wrap calls to `RCTDevSettings`
- RCTBatchedBridge would subscribe to `kRCTDevSettingsDidUpdateNotification`
- In the NSNotification handler, if the DevSettings value changed for either `kRCTDevSettingLiveReloadEnabled` or `kRCTDevSettingExecutorOverrideClass`, RCTBatchedBridge would reload itself
- In `initModulesWithDispatchGroup:` there would be some other logic for building an instance of the corresponding module. I think the logic would change slightly from what it is now, because we'd need to make sure the RCTDevSettings module is initialized before we know what the executor class is.

I'm not super familiar with the `RCTWebSocketObserver` code inside DevSettings, but I think it would stay mostly the same (except removing the explicit `reload` call, since it would happen anyway). Also, this strategy depends on the `NSNotification` interface which we left comments about already, so I'd like to resolve that discussion before implementing any of this.

ide commented:
The executorClass changes look like they could be their own diff, it also seems like there would be a lot of different states the bridge and devsettings could be in, I'd lean towards keeping this diff simple. We could add a comment and/or open a github issue to track this proposal in the meantime.

terribleben commented:
@javache hope you've had a restful holiday. Let me know if you've had a chance to look at this and if you have any other questions. :)

ide commented:
@javache could you take a look at this when you get another chance?

javache commented on React/Modules/RCTDevSettings.m:
```diff
@@ -0,0 +1,509 @@
+
+@end
+
+@interface RCTDevSettings () <RCTBridgeModule, RCTInvalidating, RCTWebSocketObserverDelegate>
```
`RCTWebSocketObserverDelegate` sounds like that should be split off, like RCTDevMenu, to another consumer of RCTDevSettings

javache commented on React/Modules/RCTDevSettings.m:
```diff
@@ -0,0 +1,509 @@
+  [_bridge reload];
+}
+
+RCT_EXPORT_METHOD(setIsDebuggingRemotely:(BOOL)enabled)
```
What is this used for?

javache commented:
I think I still stand by my original point: going from a strongly typed interface to a value-for-key / string-based one is a regression. This makes it harder to statically analyze our codebase, reason about callsites or even deprecate properties correctly. I'm fine with splitting RCTDevSettings from RCTDevMenu, but it should be a strongly typed interface. (It's ok for the data source to use this generic API though).

The interface with the data source needs to be more clearly defined: the way I see it:

RCTDevSettings provides all the interaction with the bridge and the rest of the app's runtime. RCTDevSettingsDataSource is the backing store for RCTDevSettings and is used to get/set values by RCTDevSettings. It should **not** alter values itself and try to update the devsettings (data flows only in one way). RCTDevMenu and our websocket (RCTDevConnection?) make use of RCTDevSettings: they get information from dev settings and write commands to it to alter state. Through that interface RCTDevSettings can decide to persist any settings using its RCTDevSettingsDataSource.

terribleben commented:
Thanks @javache, makes sense. I'll follow up next week.

terribleben commented:
Hi there! The most recent commit addresses your last feedback:
- RCTDevSettings has only a strongly typed interface, and no key/string interface.
- I completely removed the NSNotification-based updates between RCTDevSettings and its data source.

One side effect of this change is that `RCTDevMenuItem toggleItem...` can't be supported any more, since this relied on the key-based interface. I updated all occurrences of this in the codebase. This would be a breaking change if you're using it anywhere else internally.

Splitting out `RCTWebSocketObserverDelegate` sounds like a cool idea for a follow up PR. The purpose of this one is to isolate the menu UI code and to break out the settings storage mechanism. I would prefer to limit it to that for now if you're okay with it. Afterward there would be potentially lots of opportunities to clean up various individual dev tools.

terribleben commented:
Rebased again to include a few new commits that have happened in the last couple weeks. All master commits to the old Dev Menu are going to conflict with this, so it'd be good to move forward while it's still a clean merge.

javache commented on React/Modules/RCTDevSettings.h:
```diff
@@ -0,0 +1,114 @@
+ * Whether automatic polling for JS code changes is enabled. Only applicable when
+ * running the app from a server.
+ */
+@property (nonatomic, assign, setter=setLiveReloadEnabled:) BOOL isLiveReloadEnabled;
```
Why do we define these properties twice? Once a readonly and once here as writable? I'm actually not sure why they would be writable anyway, since these are all things that are derived from the bridge's execution state.

terribleben replied:
I believe you might be conflating `isLiveReloadAvailable` (whether it's possible to use live reload) with `isLiveReloadEnabled`(whether we are currently using live reload). Both properties are defined just once. Only the second property is settable.

javache commented on React/Modules/RCTDevSettings.h:
```diff
@@ -0,0 +1,114 @@
+ * Manually reload the application. Equivalent to calling [bridge reload]
+ * directly, but can be called from JS.
+ */
+- (void)reload;
```
We can expose it to JS without exposing it through the header here. Let's have One Way (tm) to do things

terribleben replied:
This is exposed in the header because RCTDevMenu's older deprecated `reload` method calls it. All the old RCTDevMenu exports are still exposed and deprecated (for now) in order to avoid breaking old workflows.

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+
+- (void)dealloc
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
```
Unnecessary now?

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+
+- (void)_reloadWithDefaults:(NSDictionary *)defaultValues
+{
+  NSDictionary *existingSettings = [[NSUserDefaults standardUserDefaults] objectForKey:kRCTDevSettingsUserDefaultsKey];
```
Perf tip: grab the `[NSUserDefaults standardUserDefaults]` object once and reference that instead of calling the `standardUserDefaults` method multiple times.

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+@interface RCTDevSettingsUserDefaultsDataSource : NSObject <RCTDevSettingsDataSource>
+
+@property (nonatomic, strong) NSMutableDictionary *settings;
+@property (nonatomic, strong) NSMutableSet *keysChanged;
```
What is `keysChanged` used for?

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+    }
+  }
+  [[NSUserDefaults standardUserDefaults] setObject:_settings forKey:kRCTDevSettingsUserDefaultsKey];
+  [[NSUserDefaults standardUserDefaults] synchronize];
```
No need to call this: https://twitter.com/catfish_man/status/674727133017587712

terribleben replied:
Interesting! TIL

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+                                               object:nil];
+
+    // Delay setup until after Bridge init
+    __weak RCTDevSettings *weakSelf = self;
```
No need for weakify here, it's fine for us to keep the object alive during a `dispatch_async(dispatch_get_main_queue()`

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)updateSettingWithValue:(id)value forKey:(NSString *)key
```
Prefix this method with an _ so it's clear it's private?

javache commented on React/Modules/RCTDevSettings.h:
```diff
@@ -0,0 +1,114 @@
+
+@property (nonatomic, readonly) BOOL isHotLoadingAvailable;
+@property (nonatomic, readonly) BOOL isLiveReloadAvailable;
+@property (nonatomic, readonly) BOOL isRemoteDebugAvailable;
```
nit: isRemoteDebuggingAvailable

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+  return [[self settingForKey:kRCTDevSettingShakeToShowDevMenu] boolValue];
+}
+
+RCT_EXPORT_METHOD(setIsDebuggingRemotely:(BOOL)enabled)
```
What calls this?

terribleben replied:
The old DevMenu exports `debugRemotely:` to JS. In the new implementation, that old method is deprecated and wraps this method. The new method is still exposed to JS.

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+
+- (void)_remoteDebugSettingDidChange
+{
+  BOOL enabled = (self.isRemoteDebugAvailable && self.isDebuggingRemotely);
```
nit: no need for ( )

javache commented on React/Modules/RCTDevSettings.mm:
```diff
@@ -0,0 +1,572 @@
+  });
+}
+
+#pragma mark - RCTWebSocketObserver
```
This is probably fine for now, but I think this should go in a separate class as well.

javache commented on React/Profiler/RCTPerfMonitor.m:
```diff
@@ -158,17 +158,20 @@ - (RCTDevMenuItem *)devMenuItem
   if (!_devMenuItem) {
     __weak __typeof__(self) weakSelf = self;
```
nit: make this `__weak RCTDevSettings devSettings = self.bridge.devSettings;` to simplify the code below.

terribleben replied:
Sounds good. Note that we also need `weakSelf` because we dispatch calls to `show` and `hide` in the block.

terribleben commented:
Addressed all of your recent feedback (or responded in comments). Thanks for taking a look!

terribleben commented:
Rebased again

terribleben commented:
Hey @javache and @hramos it's great that you guys have tried to import this, but if people keep merging conflicts, it's never going to succeed. Also the test failures are unrelated as far as I can tell. Maybe we can pick a time to go ahead with this so I don't have to keep rebasing needlessly?

javache commented:
Hey @terribleben, don't worry I'm taking care of the conflicts :) I'll try and land it today.
